README.md
clumpr
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Travis Build
Status](https://travis-ci.com/UBESP-DCTV/clumpr.svg?token=wGyFLep97LHjNKfPGjkg&branch=structures)](https://travis-ci.org/UBESP-DCTV/clumpr)
[![Codecov
Status](https://codecov.io/gh/UBESP-DCTV/clumpr/branch/structures/graph/badge.svg?token=IY02gbLUth)](https://codecov.io/gh/UBESP-DCTV/clumpr)
[![Build
status](https://ci.appveyor.com/api/projects/status/2q6ylqbu14cu43am?svg=true)](https://ci.appveyor.com/project/CorradoLanera/clumpr)
[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN
status](https://www.r-pkg.org/badges/version/clumpr)](https://cran.r-project.org/package=clumpr)

**C**urrent transp**L**ant s**U**rplus **M**anagement **P**rotocol in
**R**

Last update: 2018-06-11

## Description

The `clumpr` package aims to provide a toolbox of function for model,
validate and visualize dynamics assessed by the “Protocollo Nazionale
per la Gestione delle Eccedenze di Tutti i Programmi di Trapianto” in
Italy.

The first implementation is focused on the lung-transplant centers
management.

## Example

This is a basic example which shows you how it works:

``` r
library(clumpr)

# setup centers and regions
pavia   <- center('Pavia',   'Lombardia', offered = 11, p_accept = 0.8)
bergamo <- center('Bergamo', 'Lombardia', offered =  7, p_accept = 0.5)
milano  <- center('Milano',  'Lombardia', offered =  3)

lombardia <- region(set_centers(pavia, bergamo, milano), default_p = 0.2)

# show data
lombardia
#>     Region  : Lombardia (Italy)
#>     Centers : Pavia, Bergamo, Milano (#3)
#>     Acceptance rate : 0.92 (at least one center)
#>     Offered organs  : 21 (from all the centers)
get_centers(lombardia)
#>     Center           : Pavia (Lombardia --- Italy)
#>     Acceptance rate  : 0.8
#>     Offered organs   : 11
#> 
#>     Center           : Bergamo (Lombardia --- Italy)
#>     Acceptance rate  : 0.5
#>     Offered organs   : 7
#> 
#>     Center           : Milano (Lombardia --- Italy)
#>     Acceptance rate  : <inherit from the region rate>
#>     Offered organs   : 3

# setup macroregions
padova <- center('Padova', 'Veneto', 10, 0.9)
veneto <- region(set_centers(padova))

nitp <- macroregion('NITp', set_regions(lombardia, veneto),
  initial_strip = c('lombardia',  'lombardia', 'veneto'))

# show data
nitp
#>     Macroregion : NITp (Italy)
#>     Regions     : Lombardia; Veneto (#2)
#>     Centers     : Pavia, Bergamo, Milano; Padova (#4)
#>     Acceptance rate : 0.992 (at least one center in some region)
#>     Offered organs  : 31 (from every the centers of every region)
#>     Initail strip   : Lombardia --> Lombardia --> Veneto
#>     Current strip   : Lombardia --> Lombardia --> Veneto
#>     Time period     : 0
get_regions(nitp)
#>     Region  : Lombardia (Italy)
#>     Centers : Pavia, Bergamo, Milano (#3)
#>     Acceptance rate : 0.92 (at least one center)
#>     Offered organs  : 21 (from all the centers)
#> 
#>     Region  : Veneto (Italy)
#>     Centers : Padova (#1)
#>     Acceptance rate : 0.9 (at least one center)
#>     Offered organs  : 10 (from all the centers)
get_centers(nitp)
#> $lombardia
#>     Center           : Pavia (Lombardia --- Italy)
#>     Acceptance rate  : 0.8
#>     Offered organs   : 11
#> 
#>     Center           : Bergamo (Lombardia --- Italy)
#>     Acceptance rate  : 0.5
#>     Offered organs   : 7
#> 
#>     Center           : Milano (Lombardia --- Italy)
#>     Acceptance rate  : <inherit from the region rate>
#>     Offered organs   : 3
#> 
#> 
#> $veneto
#>     Center           : Padova (Veneto --- Italy)
#>     Acceptance rate  : 0.9
#>     Offered organs   : 10

# add another region and setup a macroarea
torino   <- center('Torino', 'Piemonte', 7, 0.6)
piemonte <- region(set_centers(torino))

nord <- macroarea('Nord',
  macroregions = set_macroregions(piemonte, nitp)
)

nord
#>     Macroarea    : Nord (Italy)
#>     Macroregions : Piemonte; NITp (#2)
#>     Regions      : Piemonte; Lombardia, Veneto (#3)
#>     Centers      : Torino; Pavia, Bergamo, Milano; Padova (#5)
#>     Acceptance rate : 0.9968 (at least one center in some region)
#>     Offered organs  : 38 (from every centers of every region)
#>     Initail strip   : Piemonte --> Nitp
#>     Current strip   : Piemonte --> Nitp
#>     Time period     : 0
get_macroregions(nord)
#>     Region  : Piemonte (Italy)
#>     Centers : Torino (#1)
#>     Acceptance rate : 0.6 (at least one center)
#>     Offered organs  : 7 (from all the centers)
#> 
#>     Macroregion : NITp (Italy)
#>     Regions     : Lombardia; Veneto (#2)
#>     Centers     : Pavia, Bergamo, Milano; Padova (#4)
#>     Acceptance rate : 0.992 (at least one center in some region)
#>     Offered organs  : 31 (from every the centers of every region)
#>     Initail strip   : Lombardia --> Lombardia --> Veneto
#>     Current strip   : Lombardia --> Lombardia --> Veneto
#>     Time period     : 0
get_regions(nord)
#>     Region  : Piemonte (Italy)
#>     Centers : Torino (#1)
#>     Acceptance rate : 0.6 (at least one center)
#>     Offered organs  : 7 (from all the centers)
#> 
#>     Region  : Lombardia (Italy)
#>     Centers : Pavia, Bergamo, Milano (#3)
#>     Acceptance rate : 0.92 (at least one center)
#>     Offered organs  : 21 (from all the centers)
#> 
#>     Region  : Veneto (Italy)
#>     Centers : Padova (#1)
#>     Acceptance rate : 0.9 (at least one center)
#>     Offered organs  : 10 (from all the centers)
get_centers(nord)
#> $piemonte
#>     Center           : Torino (Piemonte --- Italy)
#>     Acceptance rate  : 0.6
#>     Offered organs   : 7
#> 
#> 
#> $lombardia
#>     Center           : Pavia (Lombardia --- Italy)
#>     Acceptance rate  : 0.8
#>     Offered organs   : 11
#> 
#>     Center           : Bergamo (Lombardia --- Italy)
#>     Acceptance rate  : 0.5
#>     Offered organs   : 7
#> 
#>     Center           : Milano (Lombardia --- Italy)
#>     Acceptance rate  : <inherit from the region rate>
#>     Offered organs   : 3
#> 
#> 
#> $veneto
#>     Center           : Padova (Veneto --- Italy)
#>     Acceptance rate  : 0.9
#>     Offered organs   : 10

roma   <- center('Roma', 'Lazio', 10, 0.9)
lazio  <- region(set_centers(roma))

sud <- macroarea('Sud',
  macroregions = set_macroregions(lazio)
)

italy <- state('Italy', set_macroareas(nord, sud))
italy
#>     State        : Italy
#>     Macroareas   : Nord; Sud (#2)
#>     Macroregions : Piemonte, NITp; Lazio (#3)
#>     Regions      : Piemonte; Lombardia, Veneto; Lazio (#4)
#>     Centers      : Torino; Pavia, Bergamo, Milano; Padova; Roma (#6)
#>     Acceptance rate : 0.99968 (at least one center in some region)
#>     Offered organs  : 48 (from every centers of every region)
#>     Initail strip   : Nord --> Sud
#>     Current strip   : Nord --> Sud
#>     Time period     : 0

get_macroareas(italy)
#>     Macroarea    : Nord (Italy)
#>     Macroregions : Piemonte; NITp (#2)
#>     Regions      : Piemonte; Lombardia, Veneto (#3)
#>     Centers      : Torino; Pavia, Bergamo, Milano; Padova (#5)
#>     Acceptance rate : 0.9968 (at least one center in some region)
#>     Offered organs  : 38 (from every centers of every region)
#>     Initail strip   : Piemonte --> Nitp
#>     Current strip   : Piemonte --> Nitp
#>     Time period     : 0
#> 
#>     Macroarea    : Sud (Italy)
#>     Macroregions : Lazio (#1)
#>     Regions      : Lazio (#1)
#>     Centers      : Roma (#1)
#>     Acceptance rate : 0.9 (at least one center in some region)
#>     Offered organs  : 10 (from every centers of every region)
#>     Initail strip   : Lazio
#>     Current strip   : Lazio
#>     Time period     : 0
get_macroregions(italy)
#>     Region  : Piemonte (Italy)
#>     Centers : Torino (#1)
#>     Acceptance rate : 0.6 (at least one center)
#>     Offered organs  : 7 (from all the centers)
#> 
#>     Macroregion : NITp (Italy)
#>     Regions     : Lombardia; Veneto (#2)
#>     Centers     : Pavia, Bergamo, Milano; Padova (#4)
#>     Acceptance rate : 0.992 (at least one center in some region)
#>     Offered organs  : 31 (from every the centers of every region)
#>     Initail strip   : Lombardia --> Lombardia --> Veneto
#>     Current strip   : Lombardia --> Lombardia --> Veneto
#>     Time period     : 0
#> 
#>     Region  : Lazio (Italy)
#>     Centers : Roma (#1)
#>     Acceptance rate : 0.9 (at least one center)
#>     Offered organs  : 10 (from all the centers)
get_regions(italy)
#>     Region  : Piemonte (Italy)
#>     Centers : Torino (#1)
#>     Acceptance rate : 0.6 (at least one center)
#>     Offered organs  : 7 (from all the centers)
#> 
#>     Region  : Lombardia (Italy)
#>     Centers : Pavia, Bergamo, Milano (#3)
#>     Acceptance rate : 0.92 (at least one center)
#>     Offered organs  : 21 (from all the centers)
#> 
#>     Region  : Veneto (Italy)
#>     Centers : Padova (#1)
#>     Acceptance rate : 0.9 (at least one center)
#>     Offered organs  : 10 (from all the centers)
#> 
#>     Region  : Lazio (Italy)
#>     Centers : Roma (#1)
#>     Acceptance rate : 0.9 (at least one center)
#>     Offered organs  : 10 (from all the centers)
get_centers(italy)
#> $piemonte
#>     Center           : Torino (Piemonte --- Italy)
#>     Acceptance rate  : 0.6
#>     Offered organs   : 7
#> 
#> 
#> $lombardia
#>     Center           : Pavia (Lombardia --- Italy)
#>     Acceptance rate  : 0.8
#>     Offered organs   : 11
#> 
#>     Center           : Bergamo (Lombardia --- Italy)
#>     Acceptance rate  : 0.5
#>     Offered organs   : 7
#> 
#>     Center           : Milano (Lombardia --- Italy)
#>     Acceptance rate  : <inherit from the region rate>
#>     Offered organs   : 3
#> 
#> 
#> $veneto
#>     Center           : Padova (Veneto --- Italy)
#>     Acceptance rate  : 0.9
#>     Offered organs   : 10
#> 
#> 
#> $lazio
#>     Center           : Roma (Lazio --- Italy)
#>     Acceptance rate  : 0.9
#>     Offered organs   : 10

# TODO
```

## Installation

You can install clumpr from GitHub with:

``` r
# install.packages("devtools")
devtools::install_github("UBESP-DCTV/clumpr")
```

## Bug reports

If you encounter a bug, please file a
[reprex](https://github.com/tidyverse/reprex) (minimal reproducible
example) to <https://github.com/UBESP-DCTV/clumpr/issues>

## Code of conduct

Please note that this project is released with a [Contributor Code of
Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree
to abide by its terms.

## References

**Protocollo Nazionale per la Gestione delle Eccedenze di Tutti i
Programmi di Trapianto**
([pdf](http://www.policlinico.mi.it/AMM/nitp/area_operatore/linee_guida/03/ProtocolloNazionaleGestioneEccedenzeCNTO140804.pdf))


NEWS.md
# clumpr 0.1.1

* Included italian data for 2015 and 2016
* Added dependency from the **svDialogs** package
* Added button to save data in RDA on the hard drive
* Interface updated with explicit number instead of sliders

# clumpr 0.1.0

* shinyapp

# clumpr 0.0.5

* Added strip support to state
* Changed strip for macroarea from region level to macroregion level
* Added tests dor state
* Updated `README`
* Added tests for state constructor.
* Added `get_macroareas()`, `get_macroregion()`, `get_region()`,
    `get_all_region()`, `get_state()`, `get_centers()`, `get_p_accept()`,
    `get_offered()`, `get_time()` methods for `state` class;
* Added `print()` method for `state`;
* Added `state()` and `set_macroareas` to create object of class `state`.

# clumpr 0.0.4

* Updated `LICENCE` file;
* Added lifecycle badge to the `README` file;
* Added `CODE_OF_CONDUCT.md` file;
* Reshaped `DESCRIPTION` accordingly to the tidy standards;
* Added support for AppVeyor.

# clumpr 0.0.3.063

* Updated `README`
* Added tests for macroareas.
* Added `get_center` generic and methods for `center` and `set_centers`
    classes;
* Added `get_p_accept()` method for `macroarea`;
* Added `get_*_strip()` (* = current or initial) generic funcitons and 
  methods for `macroarea`;
* Added `get_*_time()` (* = initial, current or final) generic and methods
  for `macroarea`;
* Added utility function `unlisit_1` to estract list components from a list
  vithout losing attributes (as happen with `unlist()`);
* Added `is_macroregion` function to check for inherit `macroregion` class;
* Added `get_macroregion()` generic function and method for `macroarea`;
* Added tests for `macroarea` and `set_macroregions`;
* Added `get_all_region` generic function and methods for `region`s,
  `macroregion`s and `set_macroregion`s, to access to the name of all the
  regions into the object;
* Added `get_regions` method for `region`s (identity function);
* Added `macroarea()` and `set_macroregions()`;
* Changed `time_period` to `current_time` in `macroregion` details and in
  the corresponding `get_time()` method;
* Added `get_region()` method for class `region`;
* Added `get_state` method for `macroarea`;
* Including `state` into regions' and macroregions' properties;
* Fixed an inconsistency in `center()`;
* Fixed mention to `viper` in `README`.

# clumpr 0.0.2.0082

* Added MS to authors;;
* Updated badges for `structures` branch;
* Added tests for correctness of output in macroregions;
* Added tests for `get_regions`, `get_centers`, `p_accept`, `set_regions`
  mehods for macroregions;
* Added tests for `print` method for `macroregion`s;
* Updatad `README`'s example;
* Added tests and check for unique region in a macroregion and unique
  centers on a region;
* Added `get_center`, `get_region`, `get_state`, `get_p_accept`,
  `get_offered`, `get_initial_strip` ,`get_current_strip`, `get_time`
  methods for `macroregion` (and generic function when needed);
* Added `print()` method for `macroregion` and `set_macroregion`;
* Added `macroregion()` to create object of class `macroregion`.

# clumpr 0.0.2.0002

* Organized `get_*()` in folder according to class.

# clumpr 0.0.1.00061

* Fixed bug of `at_least_one()` which leaded to a wrong probability
  computation;
* Fixed bug of `print.set_centers()` which didn't returned the imput object
  as invisible output;
* Imprved tests for regions.


# clumpr 0.0.1.00054

* Updated `README` with starting setup examples;
* Added `get_state()` generic, `get_state.center()` and `get_state.region()`
  methods to access to the state of a center or a region;
* Added `get_p_default()` generic and `get_p_default.region()` mothod to 
  access to the default probablity of acceptance for a center in a region 
  without an explicit one defined;
* Added `get_offered()` generic, `get_offered.center()` and
  `get_offered.region()` methods to access to the number of offered organs
  by centers and regions;
* Added `get_p_accept()` generic, `get_p_accept.center()` and
  `get_p_accept.region()` methods to access to the probability of
  acceptance for centers and regions;
* Added `print.region()` method;
* Added `get_centers.region()` method;
* Added `get_centers()` generic function to access to centers details;
* Provided storing of regions and states in lowercase while visualize them
  in Titlecase;
* Updated `regions` database in lowercase;
* Added import dependency from package `purrr`;
* Added tests for `set_centers()`;
* Added `set_centers()` to pass correct arguments to `region()`;
* Added `region()` constructor.


# clumpr 0.0.1.0007

* Added test for `zzz.R`;
* Defined generic function `get_region()` with method for class `center`;
* Reshaped badges in `README`.


# clumpr 0.0.0.9063

* Added support for Codecov
* Added support for Travis-CI
* Added dependecy from `R (>= 3.2)`
* Added tests for the constructor function `center()`;
* Provided `print.center()` methods;
* Created class `center` and its constructor function `center()`.


# clumpr 0.0.0.9006

* Added function `cat_line()` to `utils.R`;
* Added `.onAttach`, `.onLoad` `.onUnload` to `zzz.R` to provide package
  setup;
* Changed name in `clumper` according to "English" meanign.


# viper 0.0.0.9000

* Added support for `%>%` (pipe) operator to the package;
* Added a `NEWS.md` file to track changes to the package;
* Added a `README.rmd` and `README.md` files as welcome homepage to the 
  package;
* Added support fot test with `testthat` to the package;
* Provided basic structure to the package.


R/state.R
#' Constructor for object of class \code{state}
#'
#' The function \code{\link{state}()} is the constructor function for the homonymous
#' class.
#'
#' @param name [chr] the name of the state
#' @param macroareas [lst] a list of objects of class
#'        \code{\link{macroarea}}, passed by
#'        \code{\link{set_macroareas}()}. Note: all the macroarea in a
#'        state must be of the same state (obviously).
#' @param initial_strip [chr] a sequence of \code{\link{macroarea}}s'
#'        names in the \code{\link{state}} representing the order in
#'        which organs started to be distributed in the
#'        \code{\link{state}} at the beginning of the period. All the
#'        listed \code{\link{area}}s
#'        have to be included also in the \code{macroareas} field.
#'        Exculsions are permitted but repetitions are not admitted.
#'        (default is the sequence in which \code{\link{area}}s
#'        has been included in the argument \code{macroareas})
#'
#' @return An object of class \code{\link{state}}.
#' @export
#'
#' @examples
#' pavia   <- center('Pavia',   'Lombardia', offered = 5, p_accept = 0.6)
#' pavia
#'
#' bergamo <- center('Bergamo', 'Lombardia', 8)
#' milano  <- center('Milano',  'Lombardia', 10, 0.8)
#'
#' lombardia <- region(set_centers(pavia, bergamo, milano), default_p = 0.7)
#' lombardia
#'
#' padova <- center('Padova', 'Veneto', 8, 0.7)
#' veneto <- region(set_centers(padova))
#'
#' nitp <- macroregion('NITp', regions = set_regions(lombardia, veneto),
#'   initial_strip = c('lombardia', 'lombardia', 'veneto')
#' )
#' nitp
#'
#' torino   <- center('Torino', 'Piemonte', 7, 0.6)
#' piemonte <- region(set_centers(torino))
#'
#' nord <- macroarea('Nord',
#'   macroregions = set_macroregions(piemonte, nitp)
#' )
#' nord
#'
#' roma   <- center('Roma', 'Lazio', 10, 0.9)
#' lazio  <- region(set_centers(roma))
#'
#' sud <- macroarea('Sud',
#'   macroregions = set_macroregions(lazio)
#' )
#' sud
#'
#' italy <- state('Italy',
#'   macroareas = set_macroareas(nord, sud)
#' )
#' italy
state <- function(name, macroareas,
                  initial_strip = get_all_macroarea(macroareas)
) {

  # input check ---------------------------------------------------------

  ## name
  assertive::assert_is_inherited_from(name, 'character')
  assertive::assert_is_a_non_missing_nor_empty_string(name)

  ## macroareas
  if (!inherits(macroareas, 'set_macroareas')) stop(
    paste0('Argument ', crayon::silver("macroareas"),
           ' must be provide using\n',
           '       ', crayon::green('set_macroareas()'),
           ', and it isn\'t.\n\n',

           'Hint : To set `macroarea_1` and `macroarea_2` use\n',
           '       ',
           crayon::silver("set_macroregions(macroarea_1, macroarea_2)"), '.\n',
           'Note : all the macroareas must be of the\n',
           '       same states. No repetition is admitted in macroareas.'
    ),
    call. = FALSE
  )

  ## state name
  if (tolower(name) != tolower(get_state(macroareas[[1L]]))) stop(
    paste0('State ', crayon::silver("name"),
           ' must be the same as the name of the macroareas\' name',
           ', and it isn\'t:\n',
           '       name provided is   : ',  crayon::red(name), ',\n',
           '       macroareas\' name is: ', crayon::green(
             get_state(macroareas[[1L]])
           ), '.\n',
           'Note : state\'s name check is not case sensitive.'
    ),
    call. = FALSE
  )

  # constructor ----------------------------------------------------------

  p_accept_area <- purrr::map_dbl(macroareas, get_p_accept)
  total_offered <- purrr::map_dbl(macroareas, get_offered)

  structure(.Data = c(state = name),
            macroareas      = macroareas,
            acceptance_rate = at_least_one(p_accept_area),
            offered         = sum(total_offered),
            initial_strip   = initial_strip,
            current_strip   = initial_strip,
            current_time    = 0L,
            state           = get_state(macroareas[[1L]]),
            class           = 'state'
  )
}






#' List of macroarea creator
#'
#' Function used to prepare and check list of \code{\link{macroarea}}s to
#' pass to \code{macroarea} as needed to create \code{\link{state}}
#' objects.
#'
#' @param ... a set of \code{\link{macroarea}}s.
#'
#' @return The list of \code{\link{macroarea}}s provided.
#' @export
#'
#' @examples
#' pavia     <- center('Pavia',   'Lombardia', offered = 5, p_accept = 0.6)
#' bergamo   <- center('Bergamo', 'Lombardia', 8)
#' milano    <- center('Milano',  'Lombardia', 10, 0.8)
#' lombardia <- region(set_centers(pavia, bergamo, milano), default_p = 0.7)
#'
#' padova <- center('Padova', 'Veneto', 8, 0.7)
#' veneto <- region(set_centers(padova))
#'
#' nitp <- macroregion('NITp', regions = set_regions(lombardia, veneto),
#'   initial_strip = c('lombardia', 'lombardia', 'veneto')
#' )
#'
#' torino   <- center('Torino', 'Piemonte', 7, 0.6)
#' piemonte <- region(set_centers(torino))
#'
#' nord <- macroarea('Macroarea Nord',
#'   macroregions = set_macroregions(piemonte, nitp)
#' )
#'
#' roma   <- center('Roma', 'Lazio', 10, 0.9)
#' lazio  <- region(set_centers(roma))
#' sud    <- macroarea('Macroarea Sud',
#'   macroregions = set_macroregions(lazio)
#' )
#'
#' set_macroareas(nord, sud)
set_macroareas <- function(...) {
  .dots <- list(...)

  if (length(.dots) == 1L & inherits(.dots[[1]][[1]], 'macroarea')) {
    .dots <- .dots[[1]]
  }

  if (!all(purrr::map_lgl(.dots, inherits, c('macroarea')))) {
    stop(
      paste0('All ', crayon::silver("macroareas"), ' must be of class ',
             crayon::green('macroarea'), ', and they aren\'t.\n\n',
             '       Have you create all of them using ',
             crayon::silver('macroarea()'), '?'
      ),
      call. = FALSE
    )
  }

  regs <- purrr::map(.dots, get_all_region) %>%
    unlist() %>%
    unname()

  if (length(regs) != length(unique(regs))) stop(
    paste0(crayon::silver("Regions"),
      ' must be different and they aren\'t.\n\n',
      '       Regions provided are: ',
      paste(unique(regs),
        paste0(
          '(#',
          purrr::map_chr(table(regs),
            ~ if (. == 1) crayon::green(.) else crayon::red(.)
          ),
          ')'
        ),
        collapse = ', '
      ),
      '.'
    ),
    call. = FALSE
  )

  sts <- purrr::map_chr(.dots, get_state) %>%
    unique
  if (length(sts) != 1L) stop(
    paste0('All ', crayon::silver("regions"), ' must be in the same state',
           ' and they aren\'t.\n\n',

           '       States of the regions provided are: ',
           paste(crayon::red(sts), collapse = ', '), '.'
    ),
    call. = FALSE
  )

  structure(
    setNames(.dots,
      purrr::map_chr(.dots, 1L) %>%
        stringr::str_to_lower()
    ),
    class = 'set_macroareas'
  )
}


#' @inheritParams base::print
#' @describeIn set_macroareas nice (and coloured, if supported) print method.
#'
#' @export
print.set_macroareas <- function(x, ...) {
  purrr::walk(x, ~ {print(.); cat_line()})
  invisible(x)
}




#' @inheritParams base::print
#' @describeIn state nice (and coloured, if supported) print method.
#'
#' @export
print.state <- function(x, ...) {
  cat_line('    ',
    crayon::bold('State        : '),
    crayon::blue(stringr::str_to_title(x[[1]]))
  )

  stored_macroareas <- get_macroareas(x)
  cat_line('    ',
           crayon::bold('Macroareas   : '),
           purrr::map_chr(stored_macroareas, 1L) %>%
             str_to_title_if_needed() %>%
             crayon::blue() %>%
             paste(collapse = '; '),
           ' (#', length(stored_macroareas), ')'
  )


  stored_macroregions <- get_macroregions(x)
  cat_line('    ',
           crayon::bold('Macroregions : '),
           purrr::map(stored_macroareas, get_macroregions) %>%
             purrr::map(str_to_title_if_needed) %>%
             purrr::map(crayon::blue) %>%
             purrr::map(paste, collapse = ', ') %>%
             unlist %>%
             paste(collapse = '; '),
           ' (#', length(get_macroregions(x)), ')'
  )

  stored_regions <- get_macroregions(x) %>%
    purrr::map(get_regions)
  cat_line('    ',
           crayon::bold('Regions      : '),
           purrr::map(stored_regions,
                ~purrr::map_chr(.x, 1L)
           ) %>%
             purrr::map(str_to_title_if_needed) %>%
             purrr::map(crayon::blue) %>%
             purrr::map(paste, collapse = ', ') %>%
             unlist %>%
             paste(collapse = '; '),
           ' (#', length(unlist(stored_regions, recursive = FALSE)), ')'
  )

  stored_centers <- get_centers(x) %>%
    purrr::map(get_center)
  cat_line('    ',
           crayon::bold('Centers      : '),
           purrr::map(stored_centers,
                      ~ purrr::map_chr(.x, 1L)
           ) %>%
             purrr::map(stringr::str_to_title) %>%
             purrr::map(crayon::blue) %>%
             purrr::map(paste, collapse = ', ') %>%
             unlist %>%
             paste(collapse = '; '),
           ' (#', length(unlist(stored_centers, recursive = FALSE)), ')'
  )

  acc_p <- get_p_accept(x)
  if (acc_p == 0) {
    cat_line('    ',
             crayon::bold('Acceptance rate : '), crayon::red(acc_p)
    )
  } else {
    cat_line('    ',
             crayon::bold('Acceptance rate : '), crayon::green(acc_p),
             ' (at least one center in some region)'
    )
  }

  cat_line('    ',
           crayon::bold('Offered organs  : '),
           crayon::blue(get_offered(x)),
           ' (from every centers of every region)'
  )

  cat_line('    ',
           crayon::bold('Initail strip   : '),
           crayon::blue(get_initial_strip(x) %>% stringr::str_to_title()) %>%
             paste(collapse = ' --> ')
  )

  cat_line('    ',
           crayon::bold('Current strip   : '),
           crayon::blue(get_current_strip(x) %>% stringr::str_to_title()) %>%
             paste(collapse = ' --> ')
  )

  cat_line('    ',
           crayon::bold('Time period     : '),
           crayon::blue(get_time(x))
  )


  invisible(x)
}
